ql/src/test/org/apache/hadoop/hive/metastore/txn/TestTxnHandler.java

```diff
@@ -1544,21 +1546,68 @@ public void testRetryableRegex() throws Exception {
     Assert.assertTrue("regex should be retryable", result);
   }
 
-  @Test
-  public void testReplOpenTxn() throws Exception {
-    int numTxn = 50000;
+  private List<Long> openTxnForTest(int startId, int numTxn, String replPolicy) throws MetaException, NoSuchTxnException {
     conf.setIntVar(HiveConf.ConfVars.HIVE_TXN_MAX_OPEN_BATCH, numTxn);
+    int lastId = startId + numTxn;
     OpenTxnRequest rqst = new OpenTxnRequest(numTxn, "me", "localhost");
-    rqst.setReplPolicy("default.*");
-    rqst.setReplSrcTxnIds(LongStream.rangeClosed(1, numTxn)
+    rqst.setReplPolicy(replPolicy);
+    rqst.setReplSrcTxnIds(LongStream.rangeClosed(startId, lastId)
             .boxed().collect(Collectors.toList()));
     OpenTxnsResponse openedTxns = txnHandler.openTxns(rqst);
     List<Long> txnList = openedTxns.getTxn_ids();
     assertEquals(txnList.size(), numTxn);
     for (long i = 0; i < numTxn; i++) {
       long txnId = txnList.get((int) i);
-      assertEquals(i+1, txnId);
+      assertEquals(i + startId, txnId);
+    }
+    return txnList;
+  }
+
+  @Test
+  public void testReplOpenTxn() throws Exception {
+    int numTxn = 50000;
+    List<Long> txnList = openTxnForTest(1, numTxn, "default.*");
+    assert(txnList.size() == numTxn);
+    txnHandler.abortTxns(new AbortTxnsRequest(txnList));
+  }
+
+  @Test
+  public void testReplAllocWriteId() throws Exception {
+    int numTxn = 1000;
+    List<Long> txnList = openTxnForTest(1, numTxn, "default.*");
+    assert(txnList.size() == numTxn);
+
+    List<TxnToWriteId> txnToWriteId;
+    AllocateTableWriteIdsRequest allocMsg = new AllocateTableWriteIdsRequest("default", "tbl");
+    allocMsg.setTxnIds(txnList);
+    txnToWriteId = txnHandler.allocateTableWriteIds(allocMsg).getTxnToWriteIds();
+    assert txnToWriteId.size() == numTxn;
+
+    // idempotent case for default db
+    allocMsg = new AllocateTableWriteIdsRequest("default", "tbl");
+    allocMsg.setReplPolicy("default.*");
+    allocMsg.setTxnToWriteIdList(txnToWriteId);
+    List<TxnToWriteId> txnToWriteId1 = txnHandler.allocateTableWriteIds(allocMsg).getTxnToWriteIds();
+    assert (txnToWriteId.equals(txnToWriteId1));
+
+    txnHandler.abortTxns(new AbortTxnsRequest(txnList));
+    txnList = openTxnForTest(numTxn+1, numTxn, "default1.*");
+    assert(txnList.size() == numTxn);
+    txnToWriteId = new ArrayList<>();
+
+    for (int idx = 0; idx < txnList.size(); idx++) {
+      txnToWriteId.add(new TxnToWriteId(txnList.get(idx), idx+1));
     }
+    allocMsg = new AllocateTableWriteIdsRequest("default1", "tbl1");
+    allocMsg.setReplPolicy("default1.*");
+    allocMsg.setTxnToWriteIdList(txnToWriteId);
```

Comment:
Here, the input txnid should be that of source (default). Not target(default1) txn ids.

Reply:
here the transactions are started as target only ..and then those txn ids are passed to alloc the write id

Reply:
change the txn list to source